File: copy_savedvariables_to_data.ts
#!/usr/bin/env node
/** Copy AuctionExport SavedVariables into ./data with a timestamped name.
 *
 * Default WoW Retail path checked first:
 *   C:\Program Files (x86)\World of Warcraft\_retail_\WTF\Account
 *
 * The script ignores the shared "SharedVariables" folder under Account.
 */

import * as fs from "node:fs";
import * as path from "node:path";
import { stdin as input, stdout as output } from "node:process";
import * as readline from "node:readline/promises";

const DEFAULT_ACCOUNT_ROOT =
  "C:\\Program Files (x86)\\World of Warcraft\\_retail_\\WTF\\Account";

type AccountChoice = { name: string; fullPath: string };

type Cli = {
  accountRoot: string;
  account?: string;
  includeBak: boolean;
  dataDir: string;
};

function ts(): string {
  const d = new Date();
  const pad = (n: number) => String(n).padStart(2, "0");
  return `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())} ${pad(
    d.getHours()
  )}:${pad(d.getMinutes())}:${pad(d.getSeconds())}`;
}

function log(name: string, msg: string): void {
  process.stdout.write(`${ts()} [${name}] ${msg}\n`);
}

function warn(name: string, msg: string): void {
  process.stderr.write(`${ts()} [${name}] ${msg}\n`);
}

function parseArgs(argv: string[]): Cli {
  const out: Cli = {
    accountRoot: DEFAULT_ACCOUNT_ROOT,
    includeBak: false,
    dataDir: "data",
  };

  const nextValue = (i: number): [string, number] => {
    if (i + 1 >= argv.length) throw new Error(`Missing value after ${argv[i]}`);
    return [argv[i + 1], i + 1];
  };

  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === "--include-bak") {
      out.includeBak = true;
      continue;
    }
    if (a === "--account-root") {
      const [v, j] = nextValue(i);
      out.accountRoot = v;
      i = j;
      continue;
    }
    if (a === "--account") {
      const [v, j] = nextValue(i);
      out.account = v;
      i = j;
      continue;
    }
    if (a === "--data-dir") {
      const [v, j] = nextValue(i);
      out.dataDir = v;
      i = j;
      continue;
    }
    if (a.startsWith("-")) throw new Error(`Unknown argument: ${a}`);
    throw new Error(`Unexpected positional argument: ${a}`);
  }

  return out;
}

function listAccountFolders(accountRoot: string): AccountChoice[] {
  if (!fs.existsSync(accountRoot))
    throw new Error(`Account root does not exist: ${accountRoot}`);
  const st = fs.statSync(accountRoot);
  if (!st.isDirectory())
    throw new Error(`Account root is not a directory: ${accountRoot}`);

  const entries = fs.readdirSync(accountRoot, { withFileTypes: true });
  const choices: AccountChoice[] = [];

  const sorted = entries
    .filter((e) => e.isDirectory())
    .map((e) => e.name)
    .sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase()));

  for (const name of sorted) {
    const lower = name.toLowerCase();
    if (lower === "sharedvariables" || lower === "savedvariables") continue;
    choices.push({ name, fullPath: path.join(accountRoot, name) });
  }

  return choices;
}

async function promptForAccount(
  choices: AccountChoice[]
): Promise<AccountChoice> {
  if (!choices.length) throw new Error("No account folders found");
  if (choices.length === 1) return choices[0];

  log(
    "copy_savedvariables_to_data",
    "Multiple WoW account folders found. Select one:"
  );
  for (let idx = 0; idx < choices.length; idx++) {
    log("copy_savedvariables_to_data", `  ${idx + 1}) ${choices[idx].name}`);
  }

  const rl = readline.createInterface({ input, output });
  try {
    while (true) {
      const raw = (
        await rl.question(
          `${ts()} [copy_savedvariables_to_data] Enter number (or exact folder name): `
        )
      ).trim();
      if (!raw) continue;

      if (/^\d+$/.test(raw)) {
        const idx = Number.parseInt(raw, 10);
        if (idx >= 1 && idx <= choices.length) return choices[idx - 1];
        warn("copy_savedvariables_to_data", `Invalid selection: ${raw}`);
        continue;
      }

      const exact = choices.find((c) => c.name === raw);
      if (exact) return exact;
      warn("copy_savedvariables_to_data", `Unknown folder name: ${raw}`);
    }
  } finally {
    rl.close();
  }
}

async function resolveAccount(
  accountRoot: string,
  requestedAccount?: string
): Promise<AccountChoice> {
  const choices = listAccountFolders(accountRoot);

  if (requestedAccount) {
    const match = choices.find((c) => c.name === requestedAccount);
    if (match) return match;
    const names = choices.map((c) => c.name).join(", ") || "<none>";
    throw new Error(
      `Account '${requestedAccount}' not found under ${accountRoot}. Found: ${names}`
    );
  }

  return promptForAccount(choices);
}

function copyIfExists(src: string, dest: string): boolean {
  if (!fs.existsSync(src)) return false;
  fs.mkdirSync(path.dirname(dest), { recursive: true });
  fs.copyFileSync(src, dest);
  return true;
}

function timestampYmdHms(): string {
  const d = new Date();
  const pad = (n: number) => String(n).padStart(2, "0");
  return `${d.getFullYear()}${pad(d.getMonth() + 1)}${pad(d.getDate())}${pad(
    d.getHours()
  )}${pad(d.getMinutes())}${pad(d.getSeconds())}`;
}

async function main(): Promise<number> {
  let args: Cli;
  try {
    args = parseArgs(process.argv.slice(2));
  } catch (e) {
    warn("copy_savedvariables_to_data", `ERROR: ${(e as Error).message}`);
    warn(
      "copy_savedvariables_to_data",
      "Usage: node --experimental-strip-types ./copy_savedvariables_to_data.ts [--account-root PATH] [--account NAME] [--include-bak] [--data-dir data]"
    );
    return 2;
  }

  let account: AccountChoice;
  try {
    account = await resolveAccount(args.accountRoot, args.account);
  } catch (e) {
    warn("copy_savedvariables_to_data", `ERROR: ${(e as Error).message}`);
    return 2;
  }

  const savedVarsDir = path.join(account.fullPath, "SavedVariables");
  const srcLua = path.join(savedVarsDir, "AuctionExport.lua");
  const srcBak = path.join(savedVarsDir, "AuctionExport.lua.bak");

  const baseName = `AuctionExport-${timestampYmdHms()}`;
  const destDir = path.resolve(args.dataDir);
  const destLua = path.join(destDir, `${baseName}.lua`);
  const destBak = path.join(destDir, `${baseName}.lua.bak`);

  log("copy_savedvariables_to_data", `Using account: ${account.name}`);
  log("copy_savedvariables_to_data", `Source: ${savedVarsDir}`);
  log("copy_savedvariables_to_data", `Destination: ${destDir}`);

  let copiedAny = false;
  const start = performance.now();

  if (copyIfExists(srcLua, destLua)) {
    copiedAny = true;
    log("copy_savedvariables_to_data", `Copied: ${srcLua} -> ${destLua}`);
  } else {
    log("copy_savedvariables_to_data", `Not found: ${srcLua}`);
  }

  if (args.includeBak) {
    if (copyIfExists(srcBak, destBak)) {
      copiedAny = true;
      log("copy_savedvariables_to_data", `Copied: ${srcBak} -> ${destBak}`);
    } else {
      log("copy_savedvariables_to_data", `Not found: ${srcBak}`);
    }
  }

  if (!copiedAny) {
    log("copy_savedvariables_to_data", "Nothing copied.");
    return 1;
  }

  const elapsed = (performance.now() - start) / 1000;
  log("copy_savedvariables_to_data", `Done in ${elapsed.toFixed(2)}s`);
  return 0;
}

main().then((code) => {
  process.exitCode = code;
});

File: copy_savedvariables_to_data.py
#!/usr/bin/env python3
"""Copy AuctionExport SavedVariables into ./data with a timestamped name.

Default WoW Retail path checked first:
  C:\Program Files (x86)\World of Warcraft\_retail_\WTF\Account

The script ignores the shared "SharedVariables" folder under Account.

Outputs to ./data using base name:
  AuctionExport-YYYYMMDDHHMMSS

Examples:
  python .\copy_savedvariables_to_data.py
  python .\copy_savedvariables_to_data.py --include-bak
  python .\copy_savedvariables_to_data.py --account YOURACCOUNTNAME
"""

from __future__ import annotations

import argparse
import logging
import os
import shutil
import sys
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path
from typing import List, Optional


logger = logging.getLogger("copy_savedvariables_to_data")


DEFAULT_ACCOUNT_ROOT = Path(
    r"C:\Program Files (x86)\World of Warcraft\_retail_\WTF\Account"
)


@dataclass(frozen=True)
class AccountChoice:
    name: str
    path: Path


def _list_account_folders(account_root: Path) -> List[AccountChoice]:
    if not account_root.exists():
        raise FileNotFoundError(f"Account root does not exist: {account_root}")
    if not account_root.is_dir():
        raise NotADirectoryError(f"Account root is not a directory: {account_root}")

    choices: List[AccountChoice] = []
    for entry in sorted(account_root.iterdir(), key=lambda p: p.name.lower()):
        if not entry.is_dir():
            continue
        # WoW keeps some non-account folders under Account.
        if entry.name.lower() in {"sharedvariables", "savedvariables"}:
            continue
        choices.append(AccountChoice(name=entry.name, path=entry))

    return choices


def _prompt_for_account(choices: List[AccountChoice]) -> AccountChoice:
    if not choices:
        raise ValueError("No account folders found")
    if len(choices) == 1:
        return choices[0]

    logger.info("Multiple WoW account folders found. Select one:")
    for idx, c in enumerate(choices, start=1):
        logger.info(f"  {idx}) {c.name}")

    def prompt_input(msg: str) -> str:
        # Keep the input prompt timestamped too.
        prefix = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        return input(f"{prefix} [copy_savedvariables_to_data] {msg}")

    while True:
        raw = prompt_input("Enter number (or exact folder name): ").strip()
        if not raw:
            continue

        if raw.isdigit():
            idx = int(raw)
            if 1 <= idx <= len(choices):
                return choices[idx - 1]
            logger.warning(f"Invalid selection: {raw}")
            continue

        for c in choices:
            if c.name == raw:
                return c
        logger.warning(f"Unknown folder name: {raw}")


def _resolve_account(
    account_root: Path, requested_account: Optional[str]
) -> AccountChoice:
    choices = _list_account_folders(account_root)

    if requested_account:
        for c in choices:
            if c.name == requested_account:
                return c
        names = ", ".join(c.name for c in choices) or "<none>"
        raise ValueError(
            f"Account '{requested_account}' not found under {account_root}. Found: {names}"
        )

    return _prompt_for_account(choices)


def _copy_if_exists(src: Path, dest: Path) -> bool:
    if not src.exists():
        return False
    dest.parent.mkdir(parents=True, exist_ok=True)
    shutil.copy2(src, dest)
    return True


def main(argv: Optional[List[str]] = None) -> int:
    parser = argparse.ArgumentParser(
        description="Copy AuctionExport.lua (and optionally .bak) into ./data with a timestamped filename."
    )
    parser.add_argument(
        "--account-root",
        default=str(DEFAULT_ACCOUNT_ROOT),
        help=(
            "Path to WoW Retail WTF Account folder "
            "(default: C:\\Program Files (x86)\\World of Warcraft\\_retail_\\WTF\\Account)"
        ),
    )
    parser.add_argument(
        "--account",
        default=None,
        help="Account folder name under Account (if omitted, prompts when multiple exist)",
    )
    parser.add_argument(
        "--include-bak",
        action="store_true",
        help="Also copy AuctionExport.lua.bak when present",
    )
    parser.add_argument(
        "--data-dir",
        default="data",
        help="Destination directory (default: ./data)",
    )

    args = parser.parse_args(argv)

    logging.basicConfig(
        level=logging.INFO,
        format="%(asctime)s [%(name)s] %(message)s",
        datefmt="%Y-%m-%d %H:%M:%S",
    )

    account_root = Path(args.account_root)
    try:
        account = _resolve_account(account_root, args.account)
    except Exception as e:
        logger.error(f"ERROR: {e}")
        return 2

    saved_vars_dir = account.path / "SavedVariables"
    src_lua = saved_vars_dir / "AuctionExport.lua"
    src_bak = saved_vars_dir / "AuctionExport.lua.bak"

    timestamp = datetime.now().strftime("%Y%m%d%H%M%S")
    base_name = f"AuctionExport-{timestamp}"

    dest_dir = Path(args.data_dir)
    dest_lua = dest_dir / f"{base_name}.lua"
    dest_bak = dest_dir / f"{base_name}.lua.bak"

    logger.info(f"Using account: {account.name}")
    logger.info(f"Source: {saved_vars_dir}")
    logger.info(f"Destination: {dest_dir.resolve()}")

    copied_any = False
    start_total = datetime.now()

    if _copy_if_exists(src_lua, dest_lua):
        copied_any = True
        logger.info(f"Copied: {src_lua} -> {dest_lua}")
    else:
        logger.info(f"Not found: {src_lua}")

    if args.include_bak:
        if _copy_if_exists(src_bak, dest_bak):
            copied_any = True
            logger.info(f"Copied: {src_bak} -> {dest_bak}")
        else:
            logger.info(f"Not found: {src_bak}")

    if not copied_any:
        logger.info("Nothing copied.")
        return 1

    elapsed = (datetime.now() - start_total).total_seconds()
    logger.info(f"Done in {elapsed:.2f}s")

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
